### DISCUSSION TITLE
Loading Indicators on change table page

### DISCUSSION THREAD

**Discussion by michele-grifa** [filamentphp/filament#10156] [Help]
I created a table with a lot of data and if I change page, it takes some time to refresh the data.
During this time I can't see any loading indicator.

How can I show it?

**Comment by wahyusa** [comment]
I kinda have same problem, there is always a delay around 2 or more second for each action in pagination, search, and filter.
While for the first load it just takes 3 second, and when running in production mode it takes 1.8 second with cache on first load but still when executing pagination, search, and filter it takes more than 2 second.

My dev environment is below:

- Win 11 home + SSD
- Laragon
- PHP 8.1
- Laravel 10

Starting fresh new Laravel project with **"none"** starter kits then only install and use table builder.
https://filamentphp.com/docs/3.x/tables/installation#new-laravel-projects

Then generating the component
https://filamentphp.com/docs/3.x/tables/adding-a-table-to-a-livewire-component#generating-table-livewire-components-with-the-cli

Im using default Laravel 10 User model with 100 data generated from seeds.

I also tried to use `defer loading`
https://filamentphp.com/docs/3.x/tables/advanced#deferring-loading

Already disabling browser extension, running on incognito or change browser still same.

Current state:
- Kinda fast on first load
- Takes long time when using pagination, search, and filter

Expectation:
- It can load faster
- If it can't, at least show "loading" indicator, I don't see any docs about loading indicator for table.

Possible problem:
- My local dev environment or device setup incorrectly

Well, if it's all because of my device or local dev environment, having a loading indicator for each possible long request is a good addition for table builder just like DataTable .


Seems like there is already pull request for that 
https://github.com/filamentphp/filament/pull/9958

**Comment by degodfather** [comment]
Has anyone solved this issue? I'm also in the same situation. 

My table is taking a lot of time to load. When clicking the pagination button, nothing happens for a few seconds and loads the table page. The user doesn't know if the page is being loaded or stuck.

**Comment by codewithbae** [comment]
I think you can create a custom template cover that table and add a wire.loading for what you want to add a loading indicator, like this :
` <div wire:loading.class="opacity-50"
             wire:target="tableFilters,applyTableFilters,resetTableFiltersForm, nextPage, gotoPage, previousPage">
            {{ $this->table }}
            <div class="loading_indicator items-center justify-center p-4" wire:loading
                 wire:target="tableFilters,applyTableFilters,resetTableFiltersForm, nextPage, gotoPage, previousPage">
                <x-filament::loading-indicator class="h-5 w-5" />
            </div>
        </div>
`

**Comment by Umar-Farooq-Shafi** [comment]
+1

**Comment by vermaysha** [comment]
+1

**Comment by chuoke** [comment]
Hi, I need this too.

Currently, I found a temporary solution.

![image](https://github.com/user-attachments/assets/1f355682-3cb8-4cd3-b699-e5cc14e1a4e1)


First, add a CSS:

```css
.fi-ta-ctn {
    position: relative;
}
```

Second, Injection loading effect:

```php
FilamentView::registerRenderHook(
            \Filament\Tables\View\TablesRenderHook::HEADER_BEFORE,
            fn(): string => Blade::render('<div
                    wire:loading.flex
                    wire:target="tableFilters,applyTableFilters,resetTableFiltersForm,nextPage,gotoPage,previousPage"
                    class="absolute inset-0 justify-center items-center z-30 hidden bg-transparent backdrop-blur transition-all"
                    style="
                        background-image: linear-gradient(135deg, rgb(255 255 255 / var(--glass-opacity, 30%)), #0000), linear-gradient(var(--glass-reflex-degree, 100deg), rgb(255 255 255 / var(--glass-reflex-opacity, 10%)) 25%, rgb(0 0 0 / 0%) 25%);
                    "
                >
                    <x-filament::icon icon="heroicon-o-arrow-path" class="h-5 w-5 animate-spin text-primary" />
                </div>'),
        );
```

No table render before hook, if filament opens a ` \Filament\Tables\View\TablesRenderHook::TABLE_BEFORE`, it will be easy to custom.

**Reply by galliroberto** [reply]
how to apply to a change page in spa?

**Reply by chuoke** [reply]
Sorry. I'm not using SPA yet, I don't know the details.

**Reply by galliroberto** [reply]
No prob, thank you for you code, make me a boost! I added also a  filter reset.

The "problem" of spa is that use wire:navigate

**Reply by lukasinko** [reply]
@chuoke Thank you for your solution, it works perfectly. For some reason, the table needs to have header (i've workarounded it by adding `->header(new HtmlString(""))`. This + `deferLoading` option really improves the dynamic of the page.

**Comment by shaharyarhq** [comment]
You can add a simple skeleton loader to Filament tables using the following CSS + JS.

1. Add this CSS to your app.css:
```
/* Base container for each table cell */
.fi-ta-col.table-loading-overlay {
    position: relative;
    display: inline-block;
    width: 100%;
    min-height: 1.5em;
}

/* Temporarily hide real content */
.fi-ta-col.table-loading-overlay * {
    opacity: 0;
}

/* Skeleton shimmer bar */
.fi-ta-col.table-loading-overlay::after {
    content: "";
    position: absolute;
    top: 50%;
    left: 0.25rem;
    right: 0.25rem;
    height: 0.75rem;
    border-radius: 9999px;
    background-color: rgba(200, 200, 200, 0.3);
    transform: translateY(-50%);
    animation: pulse 1.5s ease-in-out infinite;
}

/* Pulse animation */
@keyframes pulse {
    0%, 100% { opacity: 0.3; }
    50% { opacity: 0.6; }
}

```

Run:

`npm run build`

2. Add this JavaScript snippet:
```
document.addEventListener("livewire:init", () => {
    Livewire.hook("commit.prepare", () => {
        document.querySelectorAll('td.fi-ta-cell .fi-ta-col').forEach(col => {
            col.classList.add('table-loading-overlay');

            const div = col.querySelector('div');
            if (div) {
                col.style.minHeight = div.offsetHeight + 'px';
            }
        });
    });

    Livewire.hook("commit", ({ succeed }) =>
        succeed(() =>
            queueMicrotask(() => {
                document.querySelectorAll('td.fi-ta-cell .fi-ta-col').forEach(col => {
                    col.classList.remove('table-loading-overlay');
                    col.style.minHeight = '';
                });
            })
        )
    );
});
```


This will apply a smooth skeleton loading effect to all Filament table cells during Livewire updates.
